File: SoundGenerator.swift
import SwiftUI
import AVFoundation
import CoreAudio // Added for some potential audio utility, though not strictly needed here.
import Accelerate // Added for math functions like sin, though often optional.

// MARK: - Waveform Type
enum WaveformType: Int, CaseIterable, Identifiable {
    case sine = 0
    case square = 1
    case sawtooth = 2
    case triangle = 3

    var id: Int { rawValue }
    var name: String {
        switch self {
        case .sine: return "Sine"
        case .square: return "Square"
        case .sawtooth: return "Sawtooth"
        case .triangle: return "Triangle"
        }
    }
}

// MARK: - Oscillator Configuration
struct OscillatorConfig {
    var type: WaveformType
    var baseFreq: Float
    var freqMod: Float
    var gain: Float
    var attack: Float
    var release: Float
}

// MARK: - Audio Engine Manager
class AudioEngineManager: ObservableObject {
    private var engine = AVAudioEngine()
    private var mixer = AVAudioMixerNode()
    private var filter = AVAudioUnitEQ(numberOfBands: 1)
    private var masterGain = AVAudioMixerNode()

    @Published var isPlaying = false
    @Published var masterVolume: Float = 0.3
    @Published var filterFrequency: Float = 2000

    @Published var oscillatorConfigs: [OscillatorConfig] = [
        OscillatorConfig(type: .sine, baseFreq: 220, freqMod: 2, gain: 0.3, attack: 0.5, release: 0.5),
        OscillatorConfig(type: .square, baseFreq: 440, freqMod: 3, gain: 0.2, attack: 0.3, release: 0.4),
        OscillatorConfig(type: .sawtooth, baseFreq: 110, freqMod: 1.5, gain: 0.25, attack: 0.4, release: 0.6),
        OscillatorConfig(type: .triangle, baseFreq: 330, freqMod: 2.5, gain: 0.28, attack: 0.2, release: 0.3)
    ]

    private var oscillatorNodes: [(player: AVAudioPlayerNode, gainNode: AVAudioMixerNode)] = []
    private var audioBuffers: [AVAudioPCMBuffer] = []
    private var frequencyTimers: [Timer] = []
    private var modulationPhase: [Double] = Array(repeating: 0.0, count: 4)
    
    // An array to hold the DispatchQueueWorkItems for Release to ensure we can cancel them
    private var releaseWorkItems: [DispatchWorkItem] = [] 
    // An array to hold work items for ongoing gain changes (slider updates)
    private var gainRampWorkItems: [DispatchWorkItem] = []


    init() {
        setupAudioEngine()
    }

    private func setupAudioEngine() {
        engine.attach(mixer)
        engine.attach(filter)
        engine.attach(masterGain)

        // Setup filter
        let filterParams = filter.bands[0]
        filterParams.filterType = .lowPass
        filterParams.frequency = filterFrequency
        filterParams.bandwidth = 1.0
        filterParams.bypass = false

        // Connect: mixer -> filter -> masterGain -> output
        engine.connect(mixer, to: filter, format: nil)
        engine.connect(filter, to: masterGain, format: nil)
        engine.connect(masterGain, to: engine.mainMixerNode, format: nil)

        masterGain.outputVolume = masterVolume

        do {
            try engine.start()
        } catch {
            print("Error starting engine: \(error.localizedDescription)")
        }
    }

    func updateMasterVolume(_ value: Float) {
        masterVolume = value
        masterGain.outputVolume = value
    }

    func updateFilterFrequency(_ value: Float) {
        filterFrequency = value
        filter.bands[0].frequency = value
    }

    func startSynthesis() {
        guard !isPlaying else { return }
        isPlaying = true
        
        // Reset modulation phase when starting
        modulationPhase = Array(repeating: 0.0, count: oscillatorConfigs.count)

        for (index, config) in oscillatorConfigs.enumerated() {
            createAndStartOscillator(config: config, index: index)
        }
    }

    private func createAndStartOscillator(config: OscillatorConfig, index: Int) {
        let playerNode = AVAudioPlayerNode()
        let gainNode = AVAudioMixerNode()

        engine.attach(playerNode)
        engine.attach(gainNode)

        let format = AVAudioFormat(standardFormatWithSampleRate: 44100, channels: 1)!

        engine.connect(playerNode, to: gainNode, format: format)
        engine.connect(gainNode, to: mixer, format: format)

        // Create audio buffer with waveform
        let buffer = createWaveformBuffer(
            frequency: config.baseFreq,
            type: config.type.rawValue,
            format: format
        )

        audioBuffers.append(buffer)
        oscillatorNodes.append((playerNode, gainNode))

        // Apply envelope
        gainNode.outputVolume = 0

        playerNode.scheduleBuffer(buffer, at: nil, options: .loops, completionHandler: nil)
        playerNode.play()

        // ⚠️ FIXED ATTACK RAMP
        let attackRamp = config.attack
        let targetGain = config.gain
        let rampSteps: Int = 50 
        let timeStep = TimeInterval(attackRamp) / TimeInterval(rampSteps)
        let volumeStep = targetGain / Float(rampSteps)

        for i in 1...rampSteps {
            DispatchQueue.main.asyncAfter(deadline: .now() + timeStep * Double(i)) {
                // Ensure volume doesn't overshoot the target gain
                gainNode.outputVolume = min(volumeStep * Float(i), targetGain)
            }
        }


        // Frequency modulation timer
        let timer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
            guard let self = self, self.isPlaying else { return }
            self.modulateFrequency(index: index)
        }
        // Using RunLoop.current.add is generally safe inside a method called from the main thread
        RunLoop.current.add(timer, forMode: .common)
        frequencyTimers.append(timer)
    }

    private func modulateFrequency(index: Int) {
        guard index < oscillatorNodes.count else { return }

        let config = oscillatorConfigs[index]
        if index < modulationPhase.count {
            // Use fmod() to keep the phase within a manageable range (e.g., 0 to 2*pi)
            modulationPhase[index] = fmod(modulationPhase[index] + 0.05, 2.0 * .pi)
        } else {
             return
        }

        let lfo = sin(modulationPhase[index])
        let freqRange = config.baseFreq * (config.freqMod - 1.0)
        let targetFreq = config.baseFreq + freqRange * Float((lfo + 1.0) / 2.0)

        // Update buffer with new frequency
        let format = AVAudioFormat(standardFormatWithSampleRate: 44100, channels: 1)!
        let newBuffer = createWaveformBuffer(frequency: targetFreq, type: config.type.rawValue, format: format)

        let (player, _) = oscillatorNodes[index]
        player.stop()
        player.scheduleBuffer(newBuffer, at: nil, options: .loops, completionHandler: nil)
        player.play()
    }

    private func createWaveformBuffer(frequency: Float, type: Int, format: AVAudioFormat) -> AVAudioPCMBuffer {
        let sampleRate = format.sampleRate
        let periodFrames = sampleRate / Double(frequency)
        let bufferSize = AVAudioFrameCount(ceil(periodFrames))
        
        let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: bufferSize)!
        buffer.frameLength = bufferSize

        let channels = buffer.floatChannelData!
        let channelData = channels[0]
        
        let twoPi: Float = 2.0 * .pi
        let phaseStep: Float = twoPi / Float(bufferSize)
        var currentPhase: Float = 0.0

        for frame in 0..<Int(bufferSize) {
            var value: Float

            switch type {
            case WaveformType.sine.rawValue: // Sine
                value = sin(currentPhase)
                currentPhase += phaseStep
            case WaveformType.square.rawValue: // Square
                value = Float(frame) < Float(bufferSize) / 2.0 ? 1.0 : -1.0
            case WaveformType.sawtooth.rawValue: // Sawtooth
                value = (2.0 * Float(frame) / Float(bufferSize - 1)) - 1.0
            case WaveformType.triangle.rawValue: // Triangle
                let normalizedFrame = Float(frame) / Float(bufferSize - 1)
                if normalizedFrame < 0.5 {
                    value = (4.0 * normalizedFrame) - 1.0
                } else {
                    value = 3.0 - (4.0 * normalizedFrame)
                }
            default:
                value = 0
            }

            channelData[frame] = value * 0.5 
        }

        return buffer
    }

    func stopSynthesis() {
        guard isPlaying else { return }
        isPlaying = false

        // Stop timers
        frequencyTimers.forEach { $0.invalidate() }
        frequencyTimers.removeAll()
        
        // Cancel any pending work items
        releaseWorkItems.forEach { $0.cancel() }
        releaseWorkItems.removeAll()
        gainRampWorkItems.forEach { $0.cancel() }
        gainRampWorkItems.removeAll()


        // ⚠️ FIXED RELEASE RAMP
        for (index, (player, gainNode)) in oscillatorNodes.enumerated() {
            let releaseRamp = oscillatorConfigs[index].release
            let currentVolume = gainNode.outputVolume
            let rampSteps: Int = 50
            let timeStep = TimeInterval(releaseRamp) / TimeInterval(rampSteps)
            let volumeStep = currentVolume / Float(rampSteps)
            
            // Dispatch the work item that handles the release ramp
            let releaseWorkItem = DispatchWorkItem {
                for i in 0...rampSteps {
                    DispatchQueue.main.asyncAfter(deadline: .now() + timeStep * Double(i)) {
                        let newVolume = currentVolume - volumeStep * Float(i)
                        gainNode.outputVolume = max(newVolume, 0)
                        
                        // Once the ramp is complete (last step), stop the nodes
                        if i == rampSteps {
                            player.stop()
                            // Detach asynchronously to avoid blocking the main thread immediately
                            DispatchQueue.main.async {
                                self.engine.detach(player)
                                self.engine.detach(gainNode)
                            }
                        }
                    }
                }
            }
            
            DispatchQueue.main.async(execute: releaseWorkItem)
            releaseWorkItems.append(releaseWorkItem)
        }

        // Clear arrays immediately after scheduling the stop/detach operations
        oscillatorNodes.removeAll()
        audioBuffers.removeAll()
        modulationPhase = Array(repeating: 0.0, count: oscillatorConfigs.count)
    }

    func updateOscillatorGain(index: Int, gain: Float) {
        guard index < oscillatorNodes.count else { return }
        
        // Cancel any previous gain ramp for this oscillator index before starting a new one
        if index < gainRampWorkItems.count {
            gainRampWorkItems[index].cancel()
        }

        oscillatorConfigs[index].gain = gain
        
        let gainNode = oscillatorNodes[index].gainNode
        let currentVolume = gainNode.outputVolume
        let targetVolume = gain
        let rampDuration: TimeInterval = 0.1 
        
        // ⚠️ FIX: Use the incremental loop for gain updates
        let rampSteps: Int = 10 
        let timeStep = rampDuration / TimeInterval(rampSteps)
        let volumeStep = (targetVolume - currentVolume) / Float(rampSteps)
        
        let gainRampWorkItem = DispatchWorkItem {
            for i in 1...rampSteps {
                DispatchQueue.main.asyncAfter(deadline: .now() + timeStep * Double(i)) {
                    let newVolume = currentVolume + volumeStep * Float(i)
                    // Ensure the volume doesn't go below 0 or above 1.0
                    gainNode.outputVolume = min(max(newVolume, 0), 1.0)
                }
            }
        }
        
        // Execute the new work item and store/update the reference
        if index >= gainRampWorkItems.count {
            gainRampWorkItems.append(gainRampWorkItem)
        } else {
            gainRampWorkItems[index] = gainRampWorkItem
        }
        DispatchQueue.main.async(execute: gainRampWorkItem)
    }
}

// MARK: - SwiftUI Views (No Changes Needed Here)
struct ContentView: View {
    @StateObject private var audioManager = AudioEngineManager()

    var body: some View {
        ScrollView {
            VStack(spacing: 20) {
                Text("🎵 Multi-Oscillator Synthesizer")
                    .font(.largeTitle)
                    .fontWeight(.bold)
                    .padding(.top)

                // Global Controls
                VStack(spacing: 15) {
                    VStack(alignment: .leading, spacing: 8) {
                        HStack {
                            Text("Master Gain")
                                .font(.headline)
                            Spacer()
                            Text(String(format: "%.2f", audioManager.masterVolume))
                                .foregroundColor(.secondary)
                        }
                        Slider(value: Binding(
                            get: { audioManager.masterVolume },
                            set: { audioManager.updateMasterVolume($0) }
                        ), in: 0...1)
                    }

                    VStack(alignment: .leading, spacing: 8) {
                        HStack {
                            Text("Low Pass Filter")
                                .font(.headline)
                            Spacer()
                            Text("\(Int(audioManager.filterFrequency)) Hz")
                                .foregroundColor(.secondary)
                        }
                        Slider(value: Binding(
                            get: { audioManager.filterFrequency },
                            set: { audioManager.updateFilterFrequency($0) }
                        ), in: 100...5000)
                    }
                }
                .padding()
                .background(Color(.systemGray6))
                .cornerRadius(15)

                // Oscillators
                ForEach(0..<audioManager.oscillatorConfigs.count, id: \.self) { index in
                    OscillatorView(
                        config: $audioManager.oscillatorConfigs[index],
                        index: index,
                        audioManager: audioManager
                    )
                }

                // Control Buttons
                VStack(spacing: 12) {
                    if !audioManager.isPlaying {
                        Button(action: {
                            audioManager.startSynthesis()
                        }) {
                            Text("Start Synthesis")
                                .font(.headline)
                                .foregroundColor(.white)
                                .frame(maxWidth: .infinity)
                                .padding()
                                .background(
                                    LinearGradient(
                                        gradient: Gradient(colors: [Color.blue, Color.purple]),
                                        startPoint: .leading,
                                        endPoint: .trailing
                                    )
                                )
                                .cornerRadius(12)
                        }
                    } else {
                        Button(action: {
                            audioManager.stopSynthesis()
                        }) {
                            Text("Stop Synthesis")
                                .font(.headline)
                                .foregroundColor(.white)
                                .frame(maxWidth: .infinity)
                                .padding()
                                .background(
                                    LinearGradient(
                                        gradient: Gradient(colors: [Color.pink, Color.red]),
                                        startPoint: .leading,
                                        endPoint: .trailing
                                    )
                                )
                                .cornerRadius(12)
                        }
                    }
                }
                .padding(.bottom, 30)
            }
            .padding()
        }
    }
}

struct OscillatorView: View {
    @Binding var config: OscillatorConfig
    let index: Int
    let audioManager: AudioEngineManager

    private var waveformName: String {
        config.type.name
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Oscillator \(index + 1) - \(waveformName)")
                .font(.headline)
                .fontWeight(.bold)

            Picker("Waveform Type", selection: $config.type) {
                ForEach(WaveformType.allCases) { type in
                    Text(type.name).tag(type)
                }
            }
            .pickerStyle(.segmented)
            .onChange(of: config.type) { _ in
                 // Note: The frequency modulation loop will automatically pick up the new waveform type on its next cycle.
            }

            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Text("Base Frequency")
                        .font(.subheadline)
                    Spacer()
                    Text("\(Int(config.baseFreq)) Hz")
                        .foregroundColor(.secondary)
                }
                Slider(value: $config.baseFreq, in: 50...880, step: 1)
            }

            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Text("Frequency Modulation")
                        .font(.subheadline)
                    Spacer()
                    Text(String(format: "%.1fx", config.freqMod))
                        .foregroundColor(.secondary)
                }
                Slider(value: $config.freqMod, in: 0.5...5, step: 0.1)
            }

            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Text("Gain")
                        .font(.subheadline)
                    Spacer()
                    Text(String(format: "%.2f", config.gain))
                        .foregroundColor(.secondary)
                }
                Slider(value: Binding(
                    get: { config.gain },
                    set: { newValue in
                        config.gain = newValue
                        audioManager.updateOscillatorGain(index: index, gain: newValue)
                    }
                ), in: 0...0.5, step: 0.01)
            }

            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Text("Attack")
                        .font(.subheadline)
                    Spacer()
                    Text(String(format: "%.1fs", config.attack))
                        .foregroundColor(.secondary)
                }
                Slider(value: $config.attack, in: 0.1...2, step: 0.1)
            }

            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Text("Release")
                        .font(.subheadline)
                    Spacer()
                    Text(String(format: "%.1fs", config.release))
                        .foregroundColor(.secondary)
                }
                Slider(value: $config.release, in: 0.1...2, step: 0.1)
            }
        }
        .padding()
        .background(Color(.systemGray6))
        .cornerRadius(15)
    }
}

File: SoundGenerator.kt
// MainActivity.kt
package com.example.synthesizer

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewmodel.compose.viewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.math.*

// MARK: - Data Classes
enum class WaveformType {
    SINE, SQUARE, SAWTOOTH, TRIANGLE
}

data class OscillatorConfig(
    var waveformType: WaveformType,
    var baseFreq: Float,
    var freqMod: Float,
    var gain: Float,
    var attack: Float,
    var release: Float
)

// MARK: - Audio Engine
class AudioEngine {
    private val sampleRate = 44100
    private val bufferSize = AudioTrack.getMinBufferSize(
        sampleRate,
        AudioFormat.CHANNEL_OUT_MONO,
        AudioFormat.ENCODING_PCM_16BIT
    )
    
    private var audioTracks = mutableListOf<AudioTrack>()
    private var isRunning = false
    private var modulationPhases = mutableListOf(0.0, 0.0, 0.0, 0.0)
    
    var masterGain = 0.3f
    var filterFrequency = 2000f
    
    fun startSynthesis(configs: List<OscillatorConfig>): List<Job> {
        isRunning = true
        modulationPhases = mutableListOf(0.0, 0.0, 0.0, 0.0)
        
        return configs.mapIndexed { index, config ->
            val audioTrack = createAudioTrack()
            audioTracks.add(audioTrack)
            audioTrack.play()
            
            kotlinx.coroutines.GlobalScope.launch(Dispatchers.Default) {
                generateAudioLoop(audioTrack, config, index)
            }
        }
    }
    
    private fun createAudioTrack(): AudioTrack {
        val audioAttributes = AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_MEDIA)
            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
            .build()
        
        val audioFormat = AudioFormat.Builder()
            .setSampleRate(sampleRate)
            .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
            .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
            .build()
        
        return AudioTrack.Builder()
            .setAudioAttributes(audioAttributes)
            .setAudioFormat(audioFormat)
            .setBufferSizeInBytes(bufferSize * 2)
            .setTransferMode(AudioTrack.MODE_STREAM)
            .build()
    }
    
    private suspend fun generateAudioLoop(
        audioTrack: AudioTrack,
        config: OscillatorConfig,
        index: Int
    ) {
        val chunkSize = 1024
        val buffer = ShortArray(chunkSize)
        var phase = 0.0
        
        while (isRunning) {
            // Calculate modulated frequency
            modulationPhases[index] += 0.01
            val lfo = sin(modulationPhases[index])
            val freqRange = config.baseFreq * (config.freqMod - 1.0f)
            val currentFreq = config.baseFreq + freqRange * ((lfo + 1.0) / 2.0).toFloat()
            
            // Generate waveform
            for (i in 0 until chunkSize) {
                val sample = generateSample(phase, config.waveformType)
                
                // Apply gain and filter
                val filtered = applyLowPassFilter(sample, currentFreq)
                val gained = filtered * config.gain * masterGain
                
                buffer[i] = (gained * 32767).toInt().coerceIn(-32768, 32767).toShort()
                
                phase += 2.0 * PI * currentFreq / sampleRate
                if (phase >= 2.0 * PI) phase -= 2.0 * PI
            }
            
            withContext(Dispatchers.IO) {
                audioTrack.write(buffer, 0, chunkSize)
            }
            
            delay(10) // Small delay to prevent overwhelming the CPU
        }
    }
    
    private fun generateSample(phase: Double, type: WaveformType): Float {
        return when (type) {
            WaveformType.SINE -> sin(phase).toFloat()
            WaveformType.SQUARE -> if (sin(phase) >= 0) 1f else -1f
            WaveformType.SAWTOOTH -> {
                val normalizedPhase = (phase / (2.0 * PI)).toFloat()
                (2f * normalizedPhase - 1f)
            }
            WaveformType.TRIANGLE -> {
                val normalizedPhase = (phase / (2.0 * PI)).toFloat()
                if (normalizedPhase < 0.5f) {
                    4f * normalizedPhase - 1f
                } else {
                    3f - 4f * normalizedPhase
                }
            }
        }
    }
    
    private var prevFilteredSample = 0f
    
    private fun applyLowPassFilter(sample: Float, frequency: Float): Float {
        val rc = 1.0f / (2.0f * PI.toFloat() * filterFrequency)
        val dt = 1.0f / sampleRate
        val alpha = dt / (rc + dt)
        
        val filtered = prevFilteredSample + alpha * (sample - prevFilteredSample)
        prevFilteredSample = filtered
        
        return filtered
    }
    
    fun stopSynthesis() {
        isRunning = false
        
        audioTracks.forEach { track ->
            try {
                track.stop()
                track.release()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        audioTracks.clear()
        prevFilteredSample = 0f
    }
    
    fun updateOscillatorGain(index: Int, gain: Float, configs: List<OscillatorConfig>) {
        if (index < configs.size) {
            configs[index].gain = gain
        }
    }
}

// MARK: - ViewModel
class SynthesizerViewModel : ViewModel() {
    private val audioEngine = AudioEngine()
    
    var isPlaying by mutableStateOf(false)
        private set
    
    var masterGain by mutableStateOf(0.3f)
    var filterFrequency by mutableStateOf(2000f)
    
    var oscillatorConfigs = mutableStateListOf(
        OscillatorConfig(WaveformType.SINE, 220f, 2f, 0.3f, 0.5f, 0.5f),
        OscillatorConfig(WaveformType.SQUARE, 440f, 3f, 0.2f, 0.3f, 0.4f),
        OscillatorConfig(WaveformType.SAWTOOTH, 110f, 1.5f, 0.25f, 0.4f, 0.6f),
        OscillatorConfig(WaveformType.TRIANGLE, 330f, 2.5f, 0.28f, 0.2f, 0.3f)
    )
    
    fun startSynthesis() {
        if (!isPlaying) {
            isPlaying = true
            audioEngine.startSynthesis(oscillatorConfigs)
        }
    }
    
    fun stopSynthesis() {
        if (isPlaying) {
            isPlaying = false
            audioEngine.stopSynthesis()
        }
    }
    
    fun updateMasterGain(value: Float) {
        masterGain = value
        audioEngine.masterGain = value
    }
    
    fun updateFilterFrequency(value: Float) {
        filterFrequency = value
        audioEngine.filterFrequency = value
    }
    
    fun updateOscillatorConfig(index: Int, config: OscillatorConfig) {
        oscillatorConfigs[index] = config
    }
    
    override fun onCleared() {
        super.onCleared()
        audioEngine.stopSynthesis()
    }
}

// MARK: - Composables
class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            SynthesizerTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    SynthesizerScreen()
                }
            }
        }
    }
}

@Composable
fun SynthesizerTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = darkColorScheme(
            primary = Color(0xFF6366F1),
            secondary = Color(0xFF8B5CF6),
            background = Color(0xFF1E1E2E),
            surface = Color(0xFF2A2A3E)
        ),
        content = content
    )
}

@Composable
fun SynthesizerScreen(viewModel: SynthesizerViewModel = viewModel()) {
    val scrollState = rememberScrollState()
    
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(
                Brush.verticalGradient(
                    colors = listOf(
                        Color(0xFF667EEA),
                        Color(0xFF764BA2)
                    )
                )
            )
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(scrollState)
                .padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(
                text = "🎵 Multi-Oscillator Synthesizer",
                fontSize = 24.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White,
                modifier = Modifier.padding(vertical = 20.dp)
            )
            
            // Global Controls
            Card(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 16.dp),
                shape = RoundedCornerShape(16.dp),
                colors = CardDefaults.cardColors(
                    containerColor = Color.White.copy(alpha = 0.15f)
                )
            ) {
                Column(modifier = Modifier.padding(16.dp)) {
                    SliderControl(
                        label = "Master Gain",
                        value = viewModel.masterGain,
                        valueRange = 0f..1f,
                        onValueChange = { viewModel.updateMasterGain(it) },
                        valueText = String.format("%.2f", viewModel.masterGain)
                    )
                    
                    Spacer(modifier = Modifier.height(16.dp))
                    
                    SliderControl(
                        label = "Low Pass Filter",
                        value = viewModel.filterFrequency,
                        valueRange = 100f..5000f,
                        onValueChange = { viewModel.updateFilterFrequency(it) },
                        valueText = "${viewModel.filterFrequency.toInt()} Hz"
                    )
                }
            }
            
            // Oscillators
            viewModel.oscillatorConfigs.forEachIndexed { index, config ->
                OscillatorCard(
                    index = index,
                    config = config,
                    onConfigChange = { viewModel.updateOscillatorConfig(index, it) }
                )
                Spacer(modifier = Modifier.height(12.dp))
            }
            
            // Control Buttons
            Spacer(modifier = Modifier.height(8.dp))
            
            if (!viewModel.isPlaying) {
                Button(
                    onClick = { viewModel.startSynthesis() },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(56.dp),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color(0xFF6366F1)
                    ),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Text("Start Synthesis", fontSize = 18.sp, fontWeight = FontWeight.Bold)
                }
            } else {
                Button(
                    onClick = { viewModel.stopSynthesis() },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(56.dp),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color(0xFFEC4899)
                    ),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Text("Stop Synthesis", fontSize = 18.sp, fontWeight = FontWeight.Bold)
                }
            }
            
            Spacer(modifier = Modifier.height(32.dp))
        }
    }
}

@Composable
fun OscillatorCard(
    index: Int,
    config: OscillatorConfig,
    onConfigChange: (OscillatorConfig) -> Unit
) {
    val waveformNames = listOf("SINE", "SQUARE", "SAWTOOTH", "TRIANGLE")
    
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(
            containerColor = Color.White.copy(alpha = 0.15f)
        )
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Text(
                text = "Oscillator ${index + 1} - ${waveformNames[index]}",
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White,
                modifier = Modifier.padding(bottom = 12.dp)
            )
            
            SliderControl(
                label = "Base Frequency",
                value = config.baseFreq,
                valueRange = 50f..880f,
                onValueChange = { onConfigChange(config.copy(baseFreq = it)) },
                valueText = "${config.baseFreq.toInt()} Hz"
            )
            
            Spacer(modifier = Modifier.height(12.dp))
            
            SliderControl(
                label = "Frequency Modulation",
                value = config.freqMod,
                valueRange = 0.5f..5f,
                onValueChange = { onConfigChange(config.copy(freqMod = it)) },
                valueText = String.format("%.1fx", config.freqMod)
            )
            
            Spacer(modifier = Modifier.height(12.dp))
            
            SliderControl(
                label = "Gain",
                value = config.gain,
                valueRange = 0f..0.5f,
                onValueChange = { onConfigChange(config.copy(gain = it)) },
                valueText = String.format("%.2f", config.gain)
            )
            
            Spacer(modifier = Modifier.height(12.dp))
            
            SliderControl(
                label = "Attack",
                value = config.attack,
                valueRange = 0.1f..2f,
                onValueChange = { onConfigChange(config.copy(attack = it)) },
                valueText = String.format("%.1fs", config.attack)
            )
            
            Spacer(modifier = Modifier.height(12.dp))
            
            SliderControl(
                label = "Release",
                value = config.release,
                valueRange = 0.1f..2f,
                onValueChange = { onConfigChange(config.copy(release = it)) },
                valueText = String.format("%.1fs", config.release)
            )
        }
    }
}

@Composable
fun SliderControl(
    label: String,
    value: Float,
    valueRange: ClosedFloatingPointRange<Float>,
    onValueChange: (Float) -> Unit,
    valueText: String
) {
    Column {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = label,
                color = Color.White.copy(alpha = 0.9f),
                fontSize = 14.sp
            )
            Text(
                text = valueText,
                color = Color.White.copy(alpha = 0.7f),
                fontSize = 14.sp,
                fontWeight = FontWeight.Bold
            )
        }
        Slider(
            value = value,
            onValueChange = onValueChange,
            valueRange = valueRange,
            colors = SliderDefaults.colors(
                thumbColor = Color.White,
                activeTrackColor = Color.White.copy(alpha = 0.7f),
                inactiveTrackColor = Color.White.copy(alpha = 0.3f)
            )
        )
    }
}